### DISCUSSION TITLE
[Support]: “database disk image is malformed” errors

### DISCUSSION THREAD

**Discussion by derekjwhitten** [blakeblackshear/frigate#9914] [General Support]
### Describe the problem you are having

Hi everyone.

I'm running Frigate 0.13.1 as an addon in Home Assistant 2024.2.1.  I have Frigate configured for 2 cameras:
* Amcrest AD410
* Dahua wired turret camera (DH-IPC-HDW3466EMP-S-AUS-BLK)

My frigate config is below.  I have two streams for each camera - one high res for recordings, one lower res for detection.  I've configured it using go2rtc and have a third camera which uses one of the high res feeds for object detection on a specific section of the frame, with most of the feed masked against motion (there is a particular part of the feed I want to detect people in but in the low res feed people aren't clear enough because it's too far away).

All has been working fine, but a couple of days ago I can no longer view clips for two of the cameras - front_gate and driveway_camera.  The garden_gate_camera is working fine (this is the one which uses the high res feed for object detction).

I upgraded to Frigate 0.13.1 the other day - I'm not sure if this is the point at which the clips stoppped working.

The Frigate logs show "database disk image is malformed".  I tried to fix this by:
* Stopping the Frigate addon
* Moving /config/frigate.db to a back-up folder
* Restarting the Frigate addon

The addon recreated the database but the problem has persisted - only clips from garden_gate_camera are working.

Accessing Home Assistant using samba share, I'm able to open the video files in VLC so the files themselves are fine, the problem seems to be with Frigate accessing its database to piece the videos back together.  Live feeds for all cameras are working fine, and I'm able to view snapshots - it's only the clips for the two cameras which aren't working.

Does anyone have any ideas how I can fix this?

### Version

0.13.1

### Frigate config file

```yaml
mqtt:
  host: 192.168.1.44
  user: mqttuser
  password: xxxxxxxxxxxxx

detectors:
  coral:
    type: edgetpu
    device: usb

birdseye:
  enabled: false

ffmpeg:
  hwaccel_args: preset-rpi-64-h264

go2rtc:
  streams:
    front_gate:
      - rtsp://admin:xxxxxxxxx@192.168.1.81:554/cam/realmonitor?channel=1&subtype=0&unicast=true
    front_gate_sub:
      - rtsp://admin:xxxxxxxxx@192.168.1.81:554/cam/realmonitor?channel=1&subtype=1&unicast=true
    driveway_camera:
      - rtsp://admin:xxxxxxxxx@192.168.1.108:554/cam/realmonitor?channel=1&subtype=0&unicast=true&proto=Onvif
    driveway_camera_sub:
      - rtsp://admin:xxxxxxxxx@192.168.1.108:554/cam/realmonitor?channel=1&subtype=2&unicast=true&proto=Onvif

cameras:
  front_gate:
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/front_gate
          roles:
            - record

        - path: rtsp://127.0.0.1:8554/front_gate_sub
          roles:
            - detect
    
    detect:
      width: 720
      height: 576
      fps: 4
      enabled: true
    
    objects:
      track:
        - person
        - car
        - bicycle
      filters:
        car:
          mask: 132,303,132,276,0,295,0,329
    
    snapshots:
      enabled: true
      timestamp: true
      bounding_box: true
      retain:
        default: 178
      required_zones:
        - front_driveway
        - neighbours_driveway
    
    record:
      enabled: true
      retain: 
        days: 7
        mode: motion
      events:
        pre_capture: 10 
        post_capture: 10 
        retain:
          default: 28
          mode: active_objects
        required_zones:   
          - front_driveway
          - neighbours_driveway

    motion:   
      mask:
        - 0,329,0,0,720,0,720,246,120,254,108,308

    zones:
      front_driveway:
        coordinates: 0,393,248,436,253,229,720,256,720,576,0,576
      neighbours_driveway:
        coordinates: 244,233,161,248,148,366,21,399,252,429
      front_driveway_left:
        coordinates: 246,576,0,576,0,393,253,434
      front_driveway_centre:
        coordinates: 509,576,493,229,250,229,245,576
      front_driveway_right:
        coordinates: 720,576,720,258,510,233,509,576

  driveway_camera:
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/driveway_camera
          roles:
            - record

        - path: rtsp://127.0.0.1:8554/driveway_camera_sub
          roles:
            - detect

    detect:
      width: 1280
      height: 720
      fps: 4
      enabled: true
    
    objects:
      track:
        - person
        - car
        - dog
      filters:
        car:
          mask:
            - 1280,720,0,720,0,0,220,0,246,489,665,475,734,0,1280,0
        person:
          threshold: 0.6
          mask:
            - 672,207,683,272,604,272,605,209
    
    snapshots:
      enabled: true
      timestamp: true
      bounding_box: true
      retain:
        default: 178
    
    record:
      enabled: true
      retain:   
        days: 14
        mode: motion
      events:
        pre_capture: 10    
        post_capture: 10   
        retain:
          default: 28
          mode: active_objects   
    
    motion:
      mask:
        - 0,0,1280,0,1280,720,320,720,477,456,970,462,1074,76,404,49,325,49,294,231,259,344,294,486,226,485,99,720,0,720
      contour_area: 20

    zones:
      garden_gate:
        coordinates: 419,154,379,159,383,58,421,55
      front_door:
        coordinates: 978,478,728,417,785,0,1076,0

  garden_gate_camera:
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/driveway_camera
          roles:
            - record

        - path: rtsp://127.0.0.1:8554/driveway_camera
          roles:
            - detect

    detect:
      width: 2688
      height: 1520
      fps: 4
      enabled: true
    
    objects:
      track:
        - person
        
    
    snapshots:
      enabled: true
      timestamp: true
      bounding_box: true
      retain:
        default: 178
    
    record:
      enabled: true
      retain: 
        days: 0
      events:
        pre_capture: 10    
        post_capture: 10   
        retain:
          default: 28
          mode: active_objects   
    
    motion:
      mask:
        - 754,0,0,0,0,1520,2688,1520,2688,0,752,0,752,103,896,113,900,324,762,334
      contour_area: 20
```

### Relevant log output

```shell
2024-02-14 09:39:31.054114840  Traceback (most recent call last):
2024-02-14 09:39:31.054117044    File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2190, in wsgi_app
2024-02-14 09:39:31.054118692      response = self.full_dispatch_request()
2024-02-14 09:39:31.054121081    File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1486, in full_dispatch_request
2024-02-14 09:39:31.054122673      rv = self.handle_user_exception(e)
2024-02-14 09:39:31.054125081    File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1484, in full_dispatch_request
2024-02-14 09:39:31.054126562      rv = self.dispatch_request()
2024-02-14 09:39:31.054128951    File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1469, in dispatch_request
2024-02-14 09:39:31.054131062      return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
2024-02-14 09:39:31.054132951    File "/opt/frigate/frigate/http.py", line 1897, in vod_event
2024-02-14 09:39:31.054134914      vod_response = vod_ts(event.camera, event.start_time, end_ts)
2024-02-14 09:39:31.054136766    File "/opt/frigate/frigate/http.py", line 1785, in vod_ts
2024-02-14 09:39:31.054138914      Recordings.select(Recordings.path, Recordings.duration, Recordings.end_time)
2024-02-14 09:39:31.054141080    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 2049, in iterator
2024-02-14 09:39:31.054142802      return iter(self.execute(database).iterator())
2024-02-14 09:39:31.054144914    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 1972, in inner
2024-02-14 09:39:31.054146654      return method(self, database, *args, **kwargs)
2024-02-14 09:39:31.054148802    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 2043, in execute
2024-02-14 09:39:31.054150284      return self._execute(database)
2024-02-14 09:39:31.054152450    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 2216, in _execute
2024-02-14 09:39:31.054153969      cursor = database.execute(self)
2024-02-14 09:39:31.054156117    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 3260, in execute
2024-02-14 09:39:31.054157691      return self.execute_sql(sql, params)
2024-02-14 09:39:31.054160061    File "/usr/local/lib/python3.9/dist-packages/playhouse/sqliteq.py", line 247, in execute_sql
2024-02-14 09:39:31.054161598      return self._execute(sql, params)
2024-02-14 09:39:31.054163802    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 3252, in execute_sql
2024-02-14 09:39:31.054165339      cursor.execute(sql, params or ())
2024-02-14 09:39:31.054167506    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 3020, in __exit__
2024-02-14 09:39:31.054169413      reraise(new_type, new_type(exc_value, *exc_args), traceback)
2024-02-14 09:39:31.054171561    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 192, in reraise
2024-02-14 09:39:31.054173043      raise value.with_traceback(tb)
2024-02-14 09:39:31.054175246    File "/usr/local/lib/python3.9/dist-packages/peewee.py", line 3252, in execute_sql
2024-02-14 09:39:31.054176783      cursor.execute(sql, params or ())
2024-02-14 09:39:31.054178561  peewee.DatabaseError: database disk image is malformed
```

### FFprobe output from your camera

```shell
front_gate
[{"return_code":1,"stderr":"","stdout":""},{"return_code":1,"stderr":"","stdout":""}]

driveway_camera
[{"return_code":1,"stderr":"","stdout":""},{"return_code":1,"stderr":"","stdout":""}]

garden_gate_camera
[{"return_code":1,"stderr":"","stdout":""},{"return_code":1,"stderr":"","stdout":""}]
```

### Operating system

HassOS

### Install method

HassOS Addon

### Coral version

USB

### Network connection

Wired

### Camera make and model

Amcrest AD410 and Dahua wired turret camera (DH-IPC-HDW3466EMP-S-AUS-BLK)

**Comment by thehellz** [comment]
Did you ever find a solution? I am in the same exact boat as you are right now.

**Reply by derekjwhitten** [reply]
Hi Matthew, I never found a solution to this specific problem.  But I did struggle with the malformed database issue for a number of months actually. Both my frigate and home assistant SqlLite DBs would become corrupt once every 1-2 weeks and I had to wipe them and restore from back-up.

I tried several solutions (a new SSD, a powered USB hub for the SSD and a stronger power supply for the Pi) but in the end I upgraded to a Mini PC over a month ago and haven't had an issue since.

I'm not sure if the issue I had with the single camera feed here was actually an issue, or just a symptom of my broader DB issues.

**Comment by lazerdave** [comment]
Hit this again — 4th time in 5 weeks on 0.17.0. Both `storage_maintainer` and `recording_cleanup` threads die silently when it happens, so nothing gets cleaned up and the recording drive fills.

Setup: M1 Mini, OrbStack 2.0.5, DB on a named Docker volume (ext4, not virtiofs — already been down that road). Latest corruption happened mid-evening during normal operation, not a shutdown or power event.

The dump/reload fix works when you catch it in time, but this last one was too far gone to dump — had to nuke and rebuild from scratch. Lost all event metadata.

The real problem is there's no recovery path. The threads crash and that's it — Frigate keeps recording happily while the disk fills to 97% because nothing is expiring old recordings anymore. Some kind of integrity check + auto-rebuild on corruption would go a long way.

**Comment by inkeep-ai-assistant** [comment]
The "database disk image is malformed" error indicates SQLite database corruption. Here's how to diagnose and fix it:

## Check Database Integrity

With Frigate stopped, run: [(1)](https://github.com/blakeblackshear/frigate/discussions/20563)

```shell
sqlite3 frigate.db "PRAGMA integrity_check;"
```
[(1)](https://github.com/blakeblackshear/frigate/discussions/20563)

## Repair the Database

If the integrity check shows errors, you can attempt to repair it: [(1)](https://github.com/blakeblackshear/frigate/discussions/20563)

```shell
sqlite3 frigate.db ".dump" > new.sql
sqlite3 new.db "PRAGMA integrity_check;"
mv frigate.db old.db
sqlite3 frigate.db < new.sql
```
[(1)](https://github.com/blakeblackshear/frigate/discussions/20563)

## If Repair Fails

If the database is too corrupted to dump, the fallback is to delete the database and recordings to start fresh. [(2)](https://github.com/blakeblackshear/frigate/discussions/17976)

## Root Cause Considerations

Based on the discussion, recurring database corruption on Raspberry Pi setups has been linked to hardware issues—specifically power delivery problems or SD card/SSD reliability. One user reported that upgrading from a Pi to a Mini PC resolved persistent corruption issues. [(3)](https://github.com/blakeblackshear/frigate/discussions/12170)

The logs showing `[Errno 30] Read-only file system` alongside the database errors suggest the underlying storage may be remounting as read-only due to filesystem errors, which would cause SQLite corruption. [(3)](https://github.com/blakeblackshear/frigate/discussions/12170)

## Silent Thread Death

When database corruption occurs, the `storage_maintainer` and `recording_cleanup` threads can die, causing recordings to stop being cleaned up and the drive to fill. There's currently no automatic recovery mechanism for this—you need to catch it manually. [(2)](https://github.com/blakeblackshear/frigate/discussions/17976)

📚 **Sources:**
1. [[Support]: Frigate database not cleaning up missing recordings after manual file deletion](https://github.com/blakeblackshear/frigate/discussions/20563)
2. [Retention configuration](https://github.com/blakeblackshear/frigate/discussions/17976)
3. [[Support]: database disk image is malformed](https://github.com/blakeblackshear/frigate/discussions/12170)



---

Have another question? Just tag [@inkeep](https://inkeep.com).
